File: bindings/java/src/main/java/com/dyber/quac100/Random.java
/*
 * Copyright © 2025 Dyber, Inc. All Rights Reserved.
 * QUAC 100 Post-Quantum Cryptographic Accelerator - Java SDK
 */
package com.dyber.quac100;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.UUID;

/**
 * Quantum Random Number Generation (QRNG) operations.
 * 
 * <p>
 * Provides true quantum random number generation using the QUAC 100's
 * integrated quantum entropy source. Random numbers generated are
 * cryptographically secure and suitable for key generation and nonces.
 * </p>
 * 
 * <h2>Example Usage:</h2>
 * 
 * <pre>{@code
 * Device device = lib.openFirstDevice();
 * Random random = device.random();
 * 
 * // Check entropy status
 * EntropyStatus status = random.getEntropyStatus();
 * System.out.println("Entropy level: " + status.getLevel() + "%");
 * 
 * // Generate random bytes
 * byte[] key = random.bytes(32);
 * 
 * // Generate random integers
 * int value = random.nextInt();
 * int inRange = random.nextInt(100); // [0, 100)
 * 
 * // Generate UUID
 * String uuid = random.uuid();
 * }</pre>
 * 
 * @author Dyber, Inc.
 * @version 1.0.0
 */
public class Random {

    private final Device device;

    // Native methods
    private static native EntropyStatus nativeGetEntropyStatus(long handle);

    private static native byte[] nativeGetBytes(long handle, int count);

    private static native int nativeGetUint32(long handle);

    private static native long nativeGetUint64(long handle);

    /**
     * Package-private constructor - instances are created by Device.
     */
    Random(Device device) {
        this.device = device;
    }

    /**
     * Gets the current entropy status.
     * 
     * @return EntropyStatus containing level, health status, etc.
     * @throws QuacException if operation fails
     */
    public EntropyStatus getEntropyStatus() throws QuacException {
        EntropyStatus status = nativeGetEntropyStatus(device.getHandle());
        if (status == null) {
            throw new QuacException(ErrorCode.INTERNAL_ERROR,
                    "Failed to get entropy status");
        }
        return status;
    }

    /**
     * Generates random bytes.
     * 
     * @param count number of bytes to generate
     * @return random byte array
     * @throws QuacException if generation fails
     */
    public byte[] bytes(int count) throws QuacException {
        if (count <= 0) {
            throw new IllegalArgumentException("Count must be positive");
        }

        byte[] result = nativeGetBytes(device.getHandle(), count);
        if (result == null || result.length != count) {
            throw new QuacException(ErrorCode.INTERNAL_ERROR,
                    "Failed to generate random bytes");
        }
        return result;
    }

    /**
     * Fills a byte array with random bytes.
     * 
     * @param buffer the buffer to fill
     * @throws QuacException if generation fails
     */
    public void nextBytes(byte[] buffer) throws QuacException {
        if (buffer == null || buffer.length == 0) {
            throw new IllegalArgumentException("Buffer cannot be null or empty");
        }

        byte[] random = bytes(buffer.length);
        System.arraycopy(random, 0, buffer, 0, buffer.length);
    }

    /**
     * Generates a random unsigned 32-bit integer.
     * 
     * @return random integer (full 32-bit range, may be negative in Java)
     * @throws QuacException if generation fails
     */
    public int nextUint32() throws QuacException {
        return nativeGetUint32(device.getHandle());
    }

    /**
     * Generates a random unsigned 64-bit integer.
     * 
     * @return random long (full 64-bit range, may be negative in Java)
     * @throws QuacException if generation fails
     */
    public long nextUint64() throws QuacException {
        return nativeGetUint64(device.getHandle());
    }

    /**
     * Generates a random non-negative integer.
     * 
     * @return random non-negative integer
     * @throws QuacException if generation fails
     */
    public int nextInt() throws QuacException {
        return nextUint32() & Integer.MAX_VALUE;
    }

    /**
     * Generates a random integer in the range [0, bound).
     * 
     * @param bound upper bound (exclusive)
     * @return random integer in [0, bound)
     * @throws QuacException if generation fails
     */
    public int nextInt(int bound) throws QuacException {
        if (bound <= 0) {
            throw new IllegalArgumentException("Bound must be positive");
        }

        // Use rejection sampling to avoid modulo bias
        int mask = bound - 1;
        if ((bound & mask) == 0) {
            // Power of 2
            return nextUint32() & mask;
        }

        // General case with rejection sampling
        int threshold = (0x80000000 - (0x80000000 % bound)) & 0x7FFFFFFF;
        while (true) {
            int r = nextUint32() & 0x7FFFFFFF;
            if (r >= threshold || (r % bound) < bound) {
                return r % bound;
            }
        }
    }

    /**
     * Generates a random integer in the range [min, max).
     * 
     * @param min lower bound (inclusive)
     * @param max upper bound (exclusive)
     * @return random integer in [min, max)
     * @throws QuacException if generation fails
     */
    public int nextInt(int min, int max) throws QuacException {
        if (min >= max) {
            throw new IllegalArgumentException("min must be less than max");
        }
        return min + nextInt(max - min);
    }

    /**
     * Generates a random non-negative long.
     * 
     * @return random non-negative long
     * @throws QuacException if generation fails
     */
    public long nextLong() throws QuacException {
        return nextUint64() & Long.MAX_VALUE;
    }

    /**
     * Generates a random long in the range [0, bound).
     * 
     * @param bound upper bound (exclusive)
     * @return random long in [0, bound)
     * @throws QuacException if generation fails
     */
    public long nextLong(long bound) throws QuacException {
        if (bound <= 0) {
            throw new IllegalArgumentException("Bound must be positive");
        }

        long mask = bound - 1;
        if ((bound & mask) == 0) {
            return nextUint64() & mask;
        }

        while (true) {
            long r = nextUint64() & Long.MAX_VALUE;
            if (r - (r % bound) + mask >= 0) {
                return r % bound;
            }
        }
    }

    /**
     * Generates a random double in the range [0.0, 1.0).
     * 
     * @return random double in [0.0, 1.0)
     * @throws QuacException if generation fails
     */
    public double nextDouble() throws QuacException {
        // Use 53 bits for double precision
        return (nextUint64() >>> 11) * (1.0 / (1L << 53));
    }

    /**
     * Generates a random double in the range [min, max).
     * 
     * @param min lower bound (inclusive)
     * @param max upper bound (exclusive)
     * @return random double in [min, max)
     * @throws QuacException if generation fails
     */
    public double nextDouble(double min, double max) throws QuacException {
        if (min >= max) {
            throw new IllegalArgumentException("min must be less than max");
        }
        return min + nextDouble() * (max - min);
    }

    /**
     * Generates a random float in the range [0.0f, 1.0f).
     * 
     * @return random float in [0.0f, 1.0f)
     * @throws QuacException if generation fails
     */
    public float nextFloat() throws QuacException {
        // Use 24 bits for float precision
        return (nextUint32() >>> 8) * (1.0f / (1 << 24));
    }

    /**
     * Generates a random boolean.
     * 
     * @return random boolean
     * @throws QuacException if generation fails
     */
    public boolean nextBoolean() throws QuacException {
        return (nextUint32() & 1) == 1;
    }

    /**
     * Generates a random UUID (version 4).
     * 
     * @return random UUID string (e.g., "550e8400-e29b-41d4-a716-446655440000")
     * @throws QuacException if generation fails
     */
    public String uuid() throws QuacException {
        byte[] randomBytes = bytes(16);

        // Set version to 4 (random UUID)
        randomBytes[6] = (byte) ((randomBytes[6] & 0x0F) | 0x40);
        // Set variant to RFC 4122
        randomBytes[8] = (byte) ((randomBytes[8] & 0x3F) | 0x80);

        // Convert to UUID format
        StringBuilder sb = new StringBuilder(36);
        for (int i = 0; i < 16; i++) {
            if (i == 4 || i == 6 || i == 8 || i == 10) {
                sb.append('-');
            }
            sb.append(String.format("%02x", randomBytes[i] & 0xFF));
        }
        return sb.toString();
    }

    /**
     * Generates a random UUID object (version 4).
     * 
     * @return random UUID
     * @throws QuacException if generation fails
     */
    public UUID nextUUID() throws QuacException {
        byte[] randomBytes = bytes(16);

        // Set version to 4 (random UUID)
        randomBytes[6] = (byte) ((randomBytes[6] & 0x0F) | 0x40);
        // Set variant to RFC 4122
        randomBytes[8] = (byte) ((randomBytes[8] & 0x3F) | 0x80);

        ByteBuffer bb = ByteBuffer.wrap(randomBytes);
        bb.order(ByteOrder.BIG_ENDIAN);
        return new UUID(bb.getLong(), bb.getLong());
    }

    /**
     * Shuffles an array in place using Fisher-Yates algorithm.
     * 
     * @param array the array to shuffle
     * @param <T>   element type
     * @throws QuacException if generation fails
     */
    public <T> void shuffle(T[] array) throws QuacException {
        if (array == null || array.length <= 1) {
            return;
        }

        for (int i = array.length - 1; i > 0; i--) {
            int j = nextInt(i + 1);
            T temp = array[i];
            array[i] = array[j];
            array[j] = temp;
        }
    }

    /**
     * Shuffles an int array in place using Fisher-Yates algorithm.
     * 
     * @param array the array to shuffle
     * @throws QuacException if generation fails
     */
    public void shuffle(int[] array) throws QuacException {
        if (array == null || array.length <= 1) {
            return;
        }

        for (int i = array.length - 1; i > 0; i--) {
            int j = nextInt(i + 1);
            int temp = array[i];
            array[i] = array[j];
            array[j] = temp;
        }
    }

    /**
     * Shuffles a byte array in place.
     * 
     * @param array the array to shuffle
     * @throws QuacException if generation fails
     */
    public void shuffle(byte[] array) throws QuacException {
        if (array == null || array.length <= 1) {
            return;
        }

        for (int i = array.length - 1; i > 0; i--) {
            int j = nextInt(i + 1);
            byte temp = array[i];
            array[i] = array[j];
            array[j] = temp;
        }
    }
}

File: bindings/csharp/Random.cs
// D:\quantacore-sdk\bindings\csharp\Random.cs
// QUAC 100 SDK - Quantum Random Number Generation
// Copyright © 2025 Dyber, Inc. All Rights Reserved.

using System.Buffers;
using System.Security.Cryptography;

namespace Dyber.Quac100;

/// <summary>
/// Quantum Random Number Generator operations
/// </summary>
public sealed class QuantumRandom : RandomNumberGenerator
{
    private readonly Device _device;
    private readonly EntropySource _preferredSource;
    private bool _disposed;

    /// <summary>Default chunk size for large requests</summary>
    public const int DefaultChunkSize = 65536; // 64KB

    /// <summary>
    /// Create QRNG operations context
    /// </summary>
    /// <param name="device">Device to use for operations</param>
    /// <param name="preferredSource">Preferred entropy source</param>
    public QuantumRandom(Device device, EntropySource preferredSource = EntropySource.Qrng)
    {
        _device = device ?? throw new ArgumentNullException(nameof(device));
        _preferredSource = preferredSource;
    }

    /// <summary>
    /// Generate random bytes
    /// </summary>
    /// <param name="count">Number of bytes to generate</param>
    /// <returns>Random byte array</returns>
    public new byte[] GetBytes(int count)
    {
        ThrowIfDisposed();

        if (count < 0)
            throw new ArgumentOutOfRangeException(nameof(count), "Count must be non-negative");

        if (count == 0)
            return Array.Empty<byte>();

        var buffer = new byte[count];
        GetBytes(buffer);
        return buffer;
    }

    /// <summary>
    /// Fill buffer with random bytes
    /// </summary>
    /// <param name="data">Buffer to fill</param>
    public override void GetBytes(byte[] data)
    {
        ThrowIfDisposed();

        if (data == null)
            throw new ArgumentNullException(nameof(data));

        GetBytes(data.AsSpan());
    }

    /// <summary>
    /// Fill span with random bytes
    /// </summary>
    public override void GetBytes(Span<byte> data)
    {
        ThrowIfDisposed();

        if (data.Length == 0)
            return;

        // For large requests, process in chunks
        if (data.Length > DefaultChunkSize)
        {
            int offset = 0;
            while (offset < data.Length)
            {
                int chunkSize = Math.Min(DefaultChunkSize, data.Length - offset);
                var chunk = data.Slice(offset, chunkSize);
                FillChunk(chunk);
                offset += chunkSize;
            }
        }
        else
        {
            FillChunk(data);
        }
    }

    private void FillChunk(Span<byte> chunk)
    {
        var buffer = chunk.ToArray();

        Quac100Status status;

        if (_preferredSource == EntropySource.Qrng)
        {
            status = (Quac100Status)NativeMethods.quac_random_bytes(
                _device.Handle, buffer, buffer.Length);
        }
        else
        {
            status = (Quac100Status)NativeMethods.quac_random_bytes_ex(
                _device.Handle, buffer, buffer.Length, (int)_preferredSource);
        }

        Quac100Exception.ThrowIfError(status, "Random byte generation failed");

        buffer.AsSpan().CopyTo(chunk);
    }

    /// <summary>
    /// Generate non-zero random bytes
    /// </summary>
    public override void GetNonZeroBytes(byte[] data)
    {
        ThrowIfDisposed();

        GetNonZeroBytes(data.AsSpan());
    }

    /// <summary>
    /// Fill span with non-zero random bytes
    /// </summary>
    public override void GetNonZeroBytes(Span<byte> data)
    {
        ThrowIfDisposed();

        // Generate random bytes and replace zeros
        GetBytes(data);

        for (int i = 0; i < data.Length; i++)
        {
            while (data[i] == 0)
            {
                var replacement = new byte[1];
                GetBytes(replacement);
                data[i] = replacement[0];
            }
        }
    }

    /// <summary>
    /// Generate a random 32-bit integer
    /// </summary>
    public int GetInt32()
    {
        Span<byte> buffer = stackalloc byte[4];
        GetBytes(buffer);
        return BitConverter.ToInt32(buffer);
    }

    /// <summary>
    /// Generate a random 32-bit integer in range [0, maxValue)
    /// </summary>
    public new int GetInt32(int maxValue)
    {
        if (maxValue <= 0)
            throw new ArgumentOutOfRangeException(nameof(maxValue), "Max value must be positive");

        return GetInt32(0, maxValue);
    }

    /// <summary>
    /// Generate a random 32-bit integer in range [minValue, maxValue)
    /// </summary>
    public new int GetInt32(int minValue, int maxValue)
    {
        if (minValue >= maxValue)
            throw new ArgumentException("minValue must be less than maxValue");

        long range = (long)maxValue - minValue;

        // Use rejection sampling for unbiased results
        long limit = uint.MaxValue - (uint.MaxValue % range);
        uint value;

        do
        {
            Span<byte> buffer = stackalloc byte[4];
            GetBytes(buffer);
            value = BitConverter.ToUInt32(buffer);
        }
        while (value >= limit);

        return (int)(minValue + (value % range));
    }

    /// <summary>
    /// Generate a random 64-bit integer
    /// </summary>
    public long GetInt64()
    {
        Span<byte> buffer = stackalloc byte[8];
        GetBytes(buffer);
        return BitConverter.ToInt64(buffer);
    }

    /// <summary>
    /// Generate a random 64-bit integer in range [0, maxValue)
    /// </summary>
    public long GetInt64(long maxValue)
    {
        if (maxValue <= 0)
            throw new ArgumentOutOfRangeException(nameof(maxValue), "Max value must be positive");

        return GetInt64(0, maxValue);
    }

    /// <summary>
    /// Generate a random 64-bit integer in range [minValue, maxValue)
    /// </summary>
    public long GetInt64(long minValue, long maxValue)
    {
        if (minValue >= maxValue)
            throw new ArgumentException("minValue must be less than maxValue");

        // Use BigInteger for large ranges if needed
        ulong range = (ulong)(maxValue - minValue);
        ulong limit = ulong.MaxValue - (ulong.MaxValue % range);
        ulong value;

        do
        {
            Span<byte> buffer = stackalloc byte[8];
            GetBytes(buffer);
            value = BitConverter.ToUInt64(buffer);
        }
        while (value >= limit);

        return minValue + (long)(value % range);
    }

    /// <summary>
    /// Generate a random double in range [0.0, 1.0)
    /// </summary>
    public double GetDouble()
    {
        // Generate 53 random bits for full double precision
        Span<byte> buffer = stackalloc byte[8];
        GetBytes(buffer);
        ulong value = BitConverter.ToUInt64(buffer);

        // Use only 53 bits for mantissa
        return (value >> 11) * (1.0 / (1UL << 53));
    }

    /// <summary>
    /// Generate a random float in range [0.0, 1.0)
    /// </summary>
    public float GetSingle()
    {
        Span<byte> buffer = stackalloc byte[4];
        GetBytes(buffer);
        uint value = BitConverter.ToUInt32(buffer);

        // Use only 24 bits for mantissa
        return (value >> 8) * (1.0f / (1U << 24));
    }

    /// <summary>
    /// Generate a random boolean
    /// </summary>
    public bool GetBoolean()
    {
        return (GetBytes(1)[0] & 1) == 1;
    }

    /// <summary>
    /// Generate a random GUID
    /// </summary>
    public Guid GetGuid()
    {
        var bytes = GetBytes(16);

        // Set version 4 (random) and variant bits
        bytes[6] = (byte)((bytes[6] & 0x0F) | 0x40); // Version 4
        bytes[8] = (byte)((bytes[8] & 0x3F) | 0x80); // Variant 1

        return new Guid(bytes);
    }

    /// <summary>
    /// Shuffle an array in place using Fisher-Yates algorithm
    /// </summary>
    public void Shuffle<T>(T[] array)
    {
        Shuffle(array.AsSpan());
    }

    /// <summary>
    /// Shuffle a span in place using Fisher-Yates algorithm
    /// </summary>
    public new void Shuffle<T>(Span<T> span)
    {
        for (int i = span.Length - 1; i > 0; i--)
        {
            int j = GetInt32(0, i + 1);
            (span[i], span[j]) = (span[j], span[i]);
        }
    }

    /// <summary>
    /// Select a random element from an array
    /// </summary>
    public T Choose<T>(T[] array)
    {
        if (array == null || array.Length == 0)
            throw new ArgumentException("Array cannot be null or empty", nameof(array));

        return array[GetInt32(0, array.Length)];
    }

    /// <summary>
    /// Select multiple random elements from an array (with replacement)
    /// </summary>
    public T[] ChooseMultiple<T>(T[] array, int count)
    {
        if (array == null || array.Length == 0)
            throw new ArgumentException("Array cannot be null or empty", nameof(array));

        var result = new T[count];
        for (int i = 0; i < count; i++)
        {
            result[i] = array[GetInt32(0, array.Length)];
        }

        return result;
    }

    /// <summary>
    /// Generate random string from character set
    /// </summary>
    public string GetString(int length, string charset = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789")
    {
        if (length < 0)
            throw new ArgumentOutOfRangeException(nameof(length));

        if (string.IsNullOrEmpty(charset))
            throw new ArgumentException("Charset cannot be null or empty", nameof(charset));

        if (length == 0)
            return string.Empty;

        var chars = new char[length];
        for (int i = 0; i < length; i++)
        {
            chars[i] = charset[GetInt32(0, charset.Length)];
        }

        return new string(chars);
    }

    /// <summary>
    /// Get entropy pool status
    /// </summary>
    public (int Level, EntropySource Source) GetEntropyStatus()
    {
        ThrowIfDisposed();

        var status = (Quac100Status)NativeMethods.quac_random_entropy_status(
            _device.Handle, out int level, out int source);

        Quac100Exception.ThrowIfError(status, "Failed to get entropy status");

        return (level, (EntropySource)source);
    }

    /// <summary>
    /// Seed the random number generator (may be ignored by hardware)
    /// </summary>
    public void Seed(ReadOnlySpan<byte> seed)
    {
        ThrowIfDisposed();

        var status = (Quac100Status)NativeMethods.quac_random_seed(
            _device.Handle, seed.ToArray(), seed.Length);

        Quac100Exception.ThrowIfError(status, "Failed to seed RNG");
    }

    /// <summary>
    /// Reseed from hardware entropy
    /// </summary>
    public void Reseed()
    {
        ThrowIfDisposed();

        var status = (Quac100Status)NativeMethods.quac_random_reseed(_device.Handle);
        Quac100Exception.ThrowIfError(status, "Failed to reseed RNG");
    }

    /// <summary>
    /// Generate random bytes asynchronously
    /// </summary>
    public Task<byte[]> GetBytesAsync(int count, CancellationToken cancellationToken = default)
    {
        return Task.Run(() => GetBytes(count), cancellationToken);
    }

    /// <summary>
    /// Benchmark random number generation
    /// </summary>
    /// <param name="bytesToGenerate">Total bytes to generate</param>
    /// <returns>Generation rate in MB/s</returns>
    public double Benchmark(int bytesToGenerate = 10 * 1024 * 1024)
    {
        ThrowIfDisposed();

        var buffer = new byte[DefaultChunkSize];
        int remaining = bytesToGenerate;

        var sw = System.Diagnostics.Stopwatch.StartNew();

        while (remaining > 0)
        {
            int toGenerate = Math.Min(buffer.Length, remaining);
            GetBytes(buffer.AsSpan(0, toGenerate));
            remaining -= toGenerate;
        }

        sw.Stop();

        return bytesToGenerate / (1024.0 * 1024.0) / sw.Elapsed.TotalSeconds;
    }

    /// <summary>
    /// Compute Shannon entropy estimate of generated bytes
    /// </summary>
    public double EstimateEntropy(int sampleSize = 100000)
    {
        ThrowIfDisposed();

        var bytes = GetBytes(sampleSize);
        var counts = new int[256];

        foreach (var b in bytes)
            counts[b]++;

        double entropy = 0;
        double n = sampleSize;

        for (int i = 0; i < 256; i++)
        {
            if (counts[i] > 0)
            {
                double p = counts[i] / n;
                entropy -= p * Math.Log2(p);
            }
        }

        return entropy;
    }

    private void ThrowIfDisposed()
    {
        if (_disposed)
            throw new ObjectDisposedException(nameof(QuantumRandom));
    }

    protected override void Dispose(bool disposing)
    {
        _disposed = true;
        base.Dispose(disposing);
    }
}

/// <summary>
/// Extension methods for random operations
/// </summary>
public static class QuantumRandomExtensions
{
    /// <summary>
    /// Create QuantumRandom from device
    /// </summary>
    public static QuantumRandom GetRandom(this Device device, EntropySource source = EntropySource.Qrng)
        => new(device, source);

    /// <summary>
    /// Fill buffer with quantum random bytes
    /// </summary>
    public static void FillRandom(this Device device, Span<byte> buffer)
    {
        using var rng = new QuantumRandom(device);
        rng.GetBytes(buffer);
    }

    /// <summary>
    /// Generate quantum random bytes
    /// </summary>
    public static byte[] GenerateRandom(this Device device, int count)
    {
        using var rng = new QuantumRandom(device);
        return rng.GetBytes(count);
    }
}